### PR TITLE
Consistent payload validation

### PR DESCRIPTION
**Pull request by jdziurzynski**
KNTK - 386
https://clkentik.atlassian.net/browse/KNTK-386

All payload validation located in /api_resources moved to requests_payload.
To make this validation consistent also refactored create/update classes in /api_resources.

### CODE REVIEW

File: kentik_api_library/kentik_api/requests_payload/populators_payload.py [hunk 1]
```diff
@@ -2,12 +2,64 @@
...
 from kentik_api.public.custom_dimension import Populator
+from kentik_api.public.errors import IncompleteObjectError
 from kentik_api.public.types import ID
 from kentik_api.requests_payload.conversions import convert, dict_from_json, from_dict
 
 # pylint: disable=too-many-instance-attributes
 
 
+@dataclass
+class PopulatorPayload:
```
Comment (danielfurman) [inline kentik_api_library/kentik_api/requests_payload/populators_payload.py:13]: We already have top level populator payload dataclass  called `_Populator`. After this change we'll have 2, which is confusing. I don't think it is a right direction. We need at least some naming convention let reader understand purpose of those 2 payload classes.

File: kentik_api_library/kentik_api/requests_payload/saved_filters_payload.py [hunk 2]
```diff
@@ -75,38 +113,45 @@ def to_saved_filters(self) -> List[SavedFilter]:
...
+    @classmethod
+    def from_custom_application(cls, saved_filter: SavedFilter):
+        check_fields(saved_filter, "Update")
+        return SavedFilterPayload.from_saved_filter(saved_filter)
+
+
 UpdateResponse = GetResponse
+
+
+def check_fields(saved_filter: SavedFilter, method: str):
```
Comment (danielfurman) [inline kentik_api_library/kentik_api/requests_payload/saved_filters_payload.py:144]:
```suggestion
def validate(saved_filter: SavedFilter, method: str):
```

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentik_api_library/kentik_api/api_resources/tags_api.py [hunk 3]
```diff
@@ -23,8 +22,6 @@ def get(self, tag_id: ID) -> Tag:
         return tags_payload.GetResponse.from_json(response.text).to_tag()
 
     def create(self, tag: Tag) -> Tag:
-        if tag.flow_tag is None:
```
[the remark this answers is not in the document]
Comment (jdziurzynski) [reply kentik_api_library/kentik_api/api_resources/tags_api.py:26]: The main purpose of the task was to move payload validation to request_payload layer. Given check was duplicated, it was already also in https://github.com/jdziurzynski/community_sdk_python/blob/06ed555b1fd3118683d39e4f97aa830732e2e8a8/kentik_api_library/kentik_api/requests_payload/tags_payload.py#L150

**Comment by mmac-m3a** [comment]
I'll wait for @danielfurman  to approve.

File: kentik_api_library/kentik_api/requests_payload/custom_dimensions_payload.py [hunk 4]
```diff
@@ -53,22 +71,43 @@ def to_custom_dimensions(self) -> List[CustomDimension]:
...
 class CreateRequest:
-    name: str
-    display_name: str
-    type: str
+
+    custom_dimension: CustomDimensionPayload
+
+    @classmethod
+    def from_custom_dimension(cls, custom_dimension: CustomDimension):
+        check_fields(custom_dimension, "Create")
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/requests_payload/custom_dimensions_payload.py]: Done

File: kentik_api_library/kentik_api/requests_payload/populators_payload.py [hunk 5]
```diff
@@ -97,78 +149,44 @@ def from_json(cls, json_string: str):
...
-    # pylint: enable=too-many-instance-attributes
-
-    def __init__(self, **kwargs) -> None:
-        self.populator = from_dict(data_class=CreateRequest._Populator, data=kwargs)
+
+    populator: PopulatorPayload
+
+    @classmethod
+    def from_populator(cls, populator: Populator):
+        check_fields(populator, "Create")
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/requests_payload/populators_payload.py]: Done

File: kentik_api_library/kentik_api/api_resources/tags_api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/api_resources/tags_api.py:26]: Ok, I didn't notice that.

File: kentik_api_library/kentik_api/requests_payload/custom_applications_payload.py [hunk 6]
```diff
@@ -2,12 +2,36 @@
...
+from kentik_api.public.errors import IncompleteObjectError
 from kentik_api.public.types import ID
 from kentik_api.requests_payload.conversions import convert, convert_or_none, dict_from_json, from_dict, list_from_json
 
 # pylint: disable=too-many-instance-attributes
 
 
+@dataclass
+class CustomApplicationPayload:
+    """This datastructure represents JSON CustomApplication payload as it is transmitted to and from KentikAPI"""
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/requests_payload/custom_applications_payload.py]: Done

File: kentik_api_library/kentik_api/requests_payload/custom_dimensions_payload.py [hunk 7]
```diff
@@ -1,12 +1,30 @@
...
 from kentik_api.public.custom_dimension import CustomDimension
+from kentik_api.public.errors import IncompleteObjectError
 from kentik_api.public.types import ID
 from kentik_api.requests_payload.conversions import convert, dict_from_json, from_dict, list_from_json
 from kentik_api.requests_payload.populators_payload import PopulatorArray
 
 
+@dataclass
+class CustomDimensionPayload:
+    """This datastructure represents JSON CustomDimension payload as it is transmitted to and from KentikAPI"""
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/requests_payload/custom_dimensions_payload.py]: Done

File: kentik_api_library/kentik_api/requests_payload/populators_payload.py [hunk 8]
```diff
@@ -2,12 +2,64 @@
...
+from kentik_api.public.errors import IncompleteObjectError
 from kentik_api.public.types import ID
 from kentik_api.requests_payload.conversions import convert, dict_from_json, from_dict
 
 # pylint: disable=too-many-instance-attributes
 
 
+@dataclass
+class PopulatorPayload:
+    """This datastructure represents JSON Populator payload as it is transmitted to and from KentikAPI"""
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/requests_payload/populators_payload.py]: Done

File: kentik_api_library/kentik_api/requests_payload/populators_payload.py (hunk 1 above)
Comment (danielfurman) [reply kentik_api_library/kentik_api/requests_payload/populators_payload.py:13]: After the changes, PopulatorPayload and PopulatorGetPayload is acceptable for me.

**Review by danielfurman — APPROVED** [review]

**Review by mmac-m3a — APPROVED** [review]
LGTM

**Pull request merged**